File: 01-java-helloworld/src/main/java/demo/DeviceCapabilities.java
package demo;

import org.protelis.lang.datatype.DeviceUID;
import org.protelis.vm.NetworkManager;
import org.protelis.vm.impl.AbstractExecutionContext;
import org.protelis.vm.impl.SimpleExecutionEnvironment;

public class DeviceCapabilities extends AbstractExecutionContext {

    private final MyDeviceUID uid;

    public DeviceCapabilities(final int uid, final NetworkManager ntmgr) {
        super(new SimpleExecutionEnvironment(), ntmgr);
        this.uid = new MyDeviceUID(uid);
    }

    public void announce(String msg) {
        System.out.println(msg);
    }

    public void announce(int a) {
        System.out.println(a);
    }

    @Override
    protected AbstractExecutionContext instance() {
        return new DeviceCapabilities(this.uid.getUID(), getNetworkManager());
    }

    @Override
    public DeviceUID getDeviceUID() {
        return uid;
    }

    @Override
    public Number getCurrentTime() {
        return System.currentTimeMillis();
    }

    @Override
    public double nextRandomDouble() {
        return Math.random();
    }
}

File: 02-kotlin-helloworld/src/main/kotlin/DeviceCapabilities.kt
package demo

import org.protelis.vm.NetworkManager
import org.protelis.vm.impl.AbstractExecutionContext
import org.protelis.vm.impl.SimpleExecutionEnvironment

class DeviceCapabilities(private val uid: Int, private val netmgr: NetworkManager) :
        AbstractExecutionContext(SimpleExecutionEnvironment(), netmgr),
        Speaker {
    private val myUID = IntDeviceUID(uid)

    override fun nextRandomDouble() = Math.random()

    override fun getDeviceUID() = myUID

    override fun getCurrentTime() =  System.currentTimeMillis()

    override fun instance() = DeviceCapabilities(uid, netmgr)

}